cip/1.accepted/CIP2015-05-13-EXISTS.adoc

```diff
@@ -4,102 +4,185 @@
+This CIP introduces existential subquery and cleans up previously existing language constructs for testing of existence.
 --
 
 toc::[]
 
 == Background
 
-In a way, Cypher already has existential subqueries - pattern predicates checking for the existence of subgraphs.
-This form is helpful, but not exhaustive in the types of subqueries that users want to be able to express.
+openCypher has two language constructs that allow testing of existence. Namely
+
+* property existence predicate: `exists( n.prop )`, which is grammarically a https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=FunctionInvocation[<FunctionInvocation>]
+* pattern predicates for the existence of subgraphs, which is grammarically a https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=RelationshipsPattern[<RelationshipsPattern>]
+
+Both are allowed as https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=Atom[<Atom>]s of expressions.
+
+Note that a <RelationshipsPattern> appearing in an expression is evaluated to a list containing the matches of the pattern.
+In a predicate context this list gets coerced to a boolean (empty list -> [underline]#*_False_*#, non-empty list -> [underline]#*_True_*#), letting this <RelationshipsPattern> behave like a predicate although it actually is a list expression.
+As a result, https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=RelationshipsPattern[<RelationshipsPattern>s] appear to behave differently depending on the context in which they appear.
+Further, https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=PatternComprehension[<PatternComprehension>] offer a more powerful and syntaxtically better separated means to   get lists of matches for a pattern within expressions.
+
+The understanding of subqueries in Cypher has evolved and Cypher gained a proper https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=NullOperatorExpression[<NullOperatorExpression>] (`IS [ NOT ] NULL`), cf. https://github.com/opencypher/openCypher/blob/master/cip/1.accepted/CIP2018-10-29-EXISTS-and-IS-NOT-NULL.adoc[CIP2018-10-29 EXISTS and IS NOT NULL].
+Both developments and the issues around <RelationshipsPattern>s suggest the need for a refinement of the language constructs that allow testing of existence.
 
 == Proposal
 
-To make this feature more powerful, this CIP suggests the addition of a new function `exists()`, and a keyword `EXISTS {}`, allowing for two different predicates:
+CIP proposes:
 
-* Property existence checking
-* Subquery existence checking
+* Restricting https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=RelationshipsPattern[<RelationshipsPattern>] to predicate contexts and boolean evaluation
```

Comment:
They are of course still valid to use in clauses such as MATCH and MERGE. Perhaps we ought to specify explicitly here that we mean within the remits of expressions?
```suggestion
* Restricting https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=RelationshipsPattern[<RelationshipsPattern>] to predicate contexts and boolean evaluation, when used in expression contexts
```

Reply:
or perhaps this is not applicable because `RelationshipsPattern` is not directly referenced in these clauses, but is only referenced by way of `Pattern`, and so there's no formal need to mention this here?

Reply:
That's right. RelationshipsPattern is only used in Atom (expressions) and PatternComprehension. So your suggestion still makes sense. :-)